pilot/pkg/model/extensions.go

```diff
@@ -75,32 +80,48 @@ func convertToWasmPluginWrapper(plugin *config.Config) *WasmPluginWrapper {
 	if u.Scheme == "" {
 		u.Scheme = ociScheme
 	}
-
+	// Normalize the image pull secret to the full resource name.
+	wasmPlugin.ImagePullSecret = toSecretResourceName(wasmPlugin.ImagePullSecret, plugin.Namespace)
 	datasource := buildDataSource(u, wasmPlugin)
-	typedConfig, err := anypb.New(&envoyWasmFilterV3.Wasm{
+	wasmExtensionConfig := &envoyWasmFilterV3.Wasm{
 		Config: &envoyExtensionsWasmV3.PluginConfig{
 			Name:          plugin.Namespace + "." + plugin.Name,
 			RootId:        wasmPlugin.PluginName,
 			Configuration: cfg,
-			Vm:            buildVMConfig(datasource, wasmPlugin.VmConfig),
+			Vm:            buildVMConfig(datasource, wasmPlugin.VmConfig, wasmPlugin.ImagePullSecret),
 		},
-	})
+	}
 	if err != nil {
 		log.Warnf("WasmPlugin %s/%s failed to marshal to TypedExtensionConfig: %s", plugin.Namespace, plugin.Name, err)
 		return nil
 	}
-	ec := &envoyCoreV3.TypedExtensionConfig{
-		Name:        plugin.Namespace + "." + plugin.Name,
-		TypedConfig: typedConfig,
-	}
 	return &WasmPluginWrapper{
-		Name:                   plugin.Name,
-		Namespace:              plugin.Namespace,
-		WasmPlugin:             *wasmPlugin,
-		ExtensionConfiguration: ec,
+		Name:                plugin.Name,
+		Namespace:           plugin.Namespace,
+		ResourceName:        plugin.Namespace + "." + plugin.Name,
+		WasmPlugin:          *wasmPlugin,
+		WasmExtensionConfig: wasmExtensionConfig,
 	}
 }
 
+// toSecretResourceName converts a imagePullSecret to a resource name referenced at Wasm SDS.
+// NOTE: the secret referenced by WasmPlugin has to be in the same namespace as the WasmPlugin,
+// so this function makes sure that the secret resource name, which will be used to retrieve secret at
+// xds generation time, has the same namespace as the WasmPlugin.
+func toSecretResourceName(name, pluginNamespace string) string {
+	if name == "" {
+		return ""
+	}
+	// If the secret already has type prefix, trim it for now.
+	name = strings.TrimPrefix(name, credentials.KubernetesSecretTypeURI)
+	// If the secret does not have namespace prefix, concat the secret name to be namespace/name.
+	if !strings.HasPrefix(name, pluginNamespace+"/") {
+		name = pluginNamespace + "/" + name
+	}
+	// Finally, add the kubernetes:// secret type.
+	return credentials.KubernetesSecretTypeURI + name
```

Comment:
instead of our own logic here, should we utilize some of the parsing in credentials/resource.go? Then it would be `resource.Namespace = pluginnamespace; return resource.String()` which feels a lot more robust.

I am very worried about the propogation of parsing logic for secret resources when any bugs will likely result in vulnerabilities